### PR TITLE
chore(react): add support for react 19

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes
Add React 19 support for React web packages. While definitely a large PR, the bulk of the changes are changes due to the removal of the global `JSX` namespace in the React 19 types.

- allow `react@^19` and `react-dom@^19` peer deps in React web and core packages (React Native does not support `react@^19` currently), example:
```diff
- "react": "^16.14.0 || ^17.0 || ^18.0"
+ "react": "^16.14 || ^17 || ^18 || ^19"
```

- update `ForwardRefPrimitive` interface to provide component `Props` and `Element` type generics to returned `React.Element` ([ref](https://github.com/aws-amplify/amplify-ui/pull/5826/files#diff-5471ceae00c4617b74b3ff3450a6a726006d166ce193519fca4cc1844f4acba5R111)). This change allows support of React 19 element typing while continuing support of React versions previous to 19.

- update `JSX.Foo` -> `React.JSX.Foo` due to global `JSX` namespace removal in React 19, example:
```diff
- function SomeComponent(): JSX.Element {}
+ function SomeComponent(): React.JSX.Element {}
```

- provide explicit `undefined` to empty `useRef` callsites: 
```diff
- useRef()
+ useRef(undefined)
```

- add `"jest"` and `"@testing-library/jest-dom"` to `"types"` field of _tsconfig.base.json_ to allow usage of `@testing-library/jest-dom` extended `jest` matchers

- update and align (where approriate) React related dev and direct dependency versions in example apps and packages:
    - `react@18.2` -> `react@^18.3`
    - `@types/react@^18.2.22` -> `@types/react@^18.3`
    - `react-dom@18.2` -> `react-dom@^18.3`
    - `@types/react-dom@^18.2.7` -> `@types/react-dom@^18.3.0`
    - `react-test-renderer@^18.2.0` -> `react-test-renderer@^18.3.0`
    - `@types/react-test-renderer@^18.0.2` -> `@types/react-test-renderer@^18.3.0`
    - `react-hook-form@"^7.43.5` -> `react-hook-form@^7.53.2`
    - `@testing-library/react-native@^12.3.0` ->`@testing-library/react-native@^12.9.0`
    - `@radix-ui/react-direction@1.0.0` -> `@radix-ui/react-direction@^1.1.0`
    - `@radix-ui/react-dropdown-menu@1.0.0` -> `@radix-ui/react-dropdown-menu@^2.1.0`
    - `@radix-ui/react-slider@1.0.0` -> `@radix-ui/react-slider@^1.2.0`
    
 - bump `StorageBrowser` size thresholds due to radix dep size increases    

- remove `"react-router-dom"` dev dependency used in a single `Link` primitive unit test and the test itself
fix typing of `DropZone` primitive components to no longer accept unsupported `forwardRef`, example ([ref](https://github.com/aws-amplify/amplify-ui/pull/5826/files#diff-726799da5f4b4b1fa2b6f6d067fdef8c45b8ba37cae6cc1e4e3f94fd49dbe337R9)):
```diff
- const Accepted: Primitive<{ children?: React.ReactNode }, 'div'> = ...
+ type AcceptedType = (props: {
+   children?: React.ReactNode;
+ }) => React.JSX.Element | null;
+ const Accepted: AcceptedType = ... 
```

- add `"@types/emscripten"` and `"@types/long"` to `@aws-amplify/ui-react-liveness` _tsconfig.json_ ([ref](https://github.com/aws-amplify/amplify-ui/pull/5826/files#diff-0323e7da2fb701dda260955531dce38c298b6b8881442a3539ba80ea4a7ec517R3))

#### Issue #, if available
closes https://github.com/aws-amplify/amplify-ui/issues/6084

#### Description of how you validated changes
- `yarn lint && yarn build`
- deployed a [sample app using React 19](https://main.d1allo2njqz3j.amplifyapp.com/) to Amplify Hosting ([source](https://github.com/calebpollman/amplify-vite-react-template))
- smoke tested React 16, 17, 18, and 19 with `vite`  and `next@15` example apps (https://github.com/calebpollman/react-19-smoke-tests) using React 19 support tagged version of `@aws-amplifyui-react@react-19` ([npm](https://www.npmjs.com/package/@aws-amplify/ui-react/v/0.0.0-react-19-0de9a74-20241217225843))

#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added
- [x] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [x] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react/src/primitives/Badge/Badge.tsx [hunk 1]
```diff
@@ -38,3 +38,6 @@ export const Badge: ForwardRefPrimitive<BaseBadgeProps, 'span'> =
   primitiveWithForwardRef(BadgePrimitive);
 
 Badge.displayName = 'Badge';
+
+type _ExFromBase = Exclude<BaseBadgeProps, BadgeProps>;
+type _ExBaseFrom = Exclude<BadgeProps, BaseBadgeProps>;
```
[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/react/src/primitives/Badge/Badge.tsx:43]: Oh good catch, was from some testing. Weird that linting didn't pick that up

[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/react/src/primitives/Badge/Badge.tsx:43]: Oh it's because I prefixed them with `_Foo` to avoid linting complaints 😅

**Review by jordanvn — DISMISSED** [review]

**Review by jordanvn — DISMISSED** [review]

File: packages/react/src/primitives/View/View.tsx [hunk 2]
```diff
@@ -17,6 +17,7 @@ const ViewPrimitive: Primitive<ViewProps, 'div'> = (
     isDisabled,
     style,
     inert,
+
```
Comment (jordanvn) [inline packages/react/src/primitives/View/View.tsx:20]: Info: is this extra line separating `...rest` intentional?

Comment (calebpollman) [inline packages/react/src/primitives/View/View.tsx:20]:
```suggestion
```

File: packages/react/src/primitives/types/view.ts [hunk 3]
```diff
@@ -94,7 +94,8 @@ interface JSXElementForwardRefExoticComponent<P = {}> {
   (props: P): React.ReactElement | null;
   defaultProps?: Partial<P> | undefined;
   displayName?: string | undefined;
-  propTypes?: React.WeakValidationMap<P> | undefined;
+  // deprecated in React@19 but needed for backwards compatibility in previous versions
+  propTypes?: any;
```
Comment (pranavosu) [inline packages/react/src/primitives/types/view.ts:98]: i think `defaultProps` was removed as well. why is only `propTypes` updated in this case?

**Review by pranavosu — APPROVED** [review]

**Review by jordanvn — APPROVED** [review]

File: packages/react/src/primitives/types/view.ts (hunk 3 above)
Comment (calebpollman) [reply packages/react/src/primitives/types/view.ts:98]: Discused offline

**Comment by morganney** [comment]
I'm still seeing older versions of dependencies that don't support React 19:

```
morganney@macos openai-realtime (feat/amplify-authn) $ npm ls react-dom
openai-realtime@1.0.0 /Users/morganney/code/openai-realtime
└─┬ ui@1.0.0 -> ./packages/ui
  ├─┬ @aws-amplify/ui-react@6.8.0
  │ ├─┬ @radix-ui/react-dropdown-menu@1.0.0
  │ │ ├─┬ @radix-ui/react-menu@1.0.0
  │ │ │ ├─┬ @radix-ui/react-dismissable-layer@1.0.0
  │ │ │ │ └── react-dom@18.3.1 deduped
  │ │ │ ├─┬ @radix-ui/react-focus-scope@1.0.0
  │ │ │ │ └── react-dom@18.3.1 deduped
  │ │ │ ├─┬ @radix-ui/react-popper@1.0.0
  │ │ │ │ ├─┬ @floating-ui/react-dom@0.7.2
  │ │ │ │ │ └── react-dom@18.3.1 deduped
  │ │ │ │ ├─┬ @radix-ui/react-arrow@1.0.0
  │ │ │ │ │ └── react-dom@18.3.1 deduped
  │ │ │ │ └── react-dom@18.3.1 deduped
  │ │ │ ├─┬ @radix-ui/react-portal@1.0.0
  │ │ │ │ └── react-dom@18.3.1 deduped
  │ │ │ ├─┬ @radix-ui/react-presence@1.0.0
  │ │ │ │ └── react-dom@18.3.1 deduped
  │ │ │ ├─┬ @radix-ui/react-roving-focus@1.0.0
  │ │ │ │ └── react-dom@18.3.1 deduped
  │ │ │ └── react-dom@18.3.1 deduped
  │ │ ├─┬ @radix-ui/react-primitive@1.0.0
  │ │ │ └── react-dom@18.3.1 deduped
  │ │ └── react-dom@18.3.1 deduped
  │ ├─┬ @radix-ui/react-slider@1.0.0
  │ │ ├─┬ @radix-ui/react-collection@1.0.0
  │ │ │ └── react-dom@18.3.1 deduped
  │ │ └── react-dom@18.3.1 deduped
  │ └── react-dom@18.3.1
  ├── react-dom@19.0.0
  └─┬ styled-components@6.1.13
    └── react-dom@18.3.1 deduped
```

When running `npm install` you'll see `npm warn ERESOLVE overriding peer dependency`.

I'm running:
```
morganney@macos openai-realtime (feat/amplify-authn) $ node -v
v22.12.0
morganney@macos openai-realtime (feat/amplify-authn) $ npm -v
10.9.0 
```

**Pull request merged**